substrate-archive/src/wasm_tracing.rs

```diff
@@ -0,0 +1,329 @@
+
+			check_target(&span.target, span.level)
+				|| wasm_target.as_ref().map(|wasm_t| check_target(wasm_t, span.level)).unwrap_or(false)
+		})
+	}
+
+	/// Formats spans based upon data types that would be more useful for querying in the context
+	/// of a relational database.
+	fn gather_span(&self, mut span: SpanMessage) -> Result<()> {
+		if span.name == WASM_TRACE_IDENTIFIER {
+			if let Some(name) = span.values.0.remove(WASM_NAME_KEY) {
+				span.name = name.to_string();
+			}
+			if let Some(target) = span.values.0.remove(WASM_TARGET_KEY) {
+				span.target = target.to_string();
+			}
+			span.file = span.values.0.remove("file").map(Into::into);
+			span.line = match span.values.0.remove("line") {
+				Some(DataType::U64(t)) => Ok(Some(t.try_into()?)),
+				None => Ok(None),
+				_ => Err(TracingError::TypeError),
+			}?;
+		}
+
+		self.span_events.lock().spans.push(span);
+		Ok(())
+	}
+}
+
+/// Stateful DataType a Tracing Value may be.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(untagged)]
+enum DataType {
+	Bool(bool),
+	I64(i64),
+	U64(u64),
+	String(String),
+}
+
+impl ToString for DataType {
+	fn to_string(&self) -> String {
+		match self {
+			DataType::Bool(b) => b.to_string(),
+			DataType::I64(i) => i.to_string(),
+			DataType::U64(u) => u.to_string(),
+			DataType::String(s) => s.to_string(),
+		}
+	}
+}
+
+impl From<DataType> for String {
+	fn from(data: DataType) -> String {
+		data.to_string()
+	}
+}
+
+#[derive(Default, Debug, Clone, Serialize, Deserialize)]
+pub struct TraceData(HashMap<String, DataType>);
+
+impl Visit for TraceData {
+	fn record_debug(&mut self, field: &Field, value: &dyn std::fmt::Debug) {
+		self.0.insert(field.name().to_string(), DataType::String(format!("{:?}", value)));
+	}
+
+	fn record_i64(&mut self, field: &Field, value: i64) {
+		self.0.insert(field.name().to_string(), DataType::I64(value));
+	}
+
+	fn record_u64(&mut self, field: &Field, value: u64) {
+		self.0.insert(field.name().to_string(), DataType::U64(value));
+	}
+
+	fn record_str(&mut self, field: &Field, value: &str) {
+		self.0.insert(field.name().to_string(), DataType::String(value.to_string()));
+	}
+
+	fn record_bool(&mut self, field: &Field, value: bool) {
+		self.0.insert(field.name().to_string(), DataType::Bool(value));
+	}
+}
+
+impl Subscriber for TraceHandler {
+	fn enabled(&self, metadata: &Metadata<'_>) -> bool {
+		self.targets.iter().any(|t| metadata.target().starts_with(t.0.as_str()))
+	}
+
+	fn new_span(&self, attrs: &Attributes<'_>) -> Id {
+		let meta = attrs.metadata();
+		let mut values = TraceData::default();
+		attrs.record(&mut values);
+
+		let id = Id::from_u64(self.counter.fetch_add(1, Ordering::Relaxed));
+		let span_message = SpanMessage {
+			id: id.clone(),
+			parent_id: attrs.parent().cloned().or_else(|| self.current_span.id()),
+			name: meta.name().to_string(),
+			target: meta.target().to_string(),
+			level: *meta.level(),
+			start_time: Utc::now(),
```

Comment:
A different approach could be to use start time (measured in ns) relative to the beginning of execution of the current block, using the monotonic `std::time::Instant` - in which case start time would be the duration between the block start and the current time. Duration would be difference between this time and the exit/close time. Perhaps this is more relevant for later analysis, but probably comes down to personal choice, so I'd be ok with either.

Reply:
Not sure I fully understand the benefit of this over just using Utc::now() -- would this give us more precise/extra information over when block execution started relative to the rest of the execution?